comms/src/protocol/rpc/server.rs

```diff
@@ -0,0 +1,431 @@
+    ) -> Result<(), RpcError>
+    {
+        if !self.executor.can_spawn() {
+            debug!(
+                target: LOG_TARGET,
+                "Closing substream to peer `{}` because maximum number of concurrent services has been reached",
+                node_id
+            );
+            framed.close().await?;
+            return Err(RpcError::MaximumConcurrencyReached);
+        }
+
+        let service = match self.service.make_service(protocol).await {
+            Ok(s) => s,
+            Err(err) => {
+                framed.close().await?;
+                return Err(err);
+            },
+        };
+
+        let version = self.perform_handshake(&mut framed).await?;
+        debug!(
+            target: LOG_TARGET,
+            "Server negotiated RPC v{} with client node `{}`", version, node_id
+        );
+
+        let service = ActivePeerRpcService {
+            config: self.config.clone(),
+            node_id,
+            framed: Some(framed),
+            service,
+            shutdown_signal: self.config.shutdown_signal.clone(),
+        };
+
+        self.executor
+            .try_spawn(service.start())
+            .map_err(|_| RpcError::MaximumConcurrencyReached)?;
+
+        Ok(())
+    }
+
+    async fn perform_handshake(&self, framed: &mut CanonicalFraming<TSubstream>) -> Result<u32, RpcError> {
+        // Only v0 is supported at this time
+        const SUPPORTED_VERSION: u32 = 0;
+        let result = time::timeout(Duration::from_secs(10), framed.next()).await;
+        match result {
+            Ok(Some(Ok(msg))) => {
+                let msg = proto::rpc::RpcSession::decode(&mut msg.freeze())?;
+                if msg.supported_versions.contains(&SUPPORTED_VERSION) {
+                    let reply = proto::rpc::RpcSessionReply {
+                        session_result: Some(proto::rpc::rpc_session_reply::SessionResult::AcceptedVersion(
+                            SUPPORTED_VERSION,
+                        )),
+                    };
+                    framed.send(reply.to_encoded_bytes().into()).await?;
+                    return Ok(SUPPORTED_VERSION);
+                }
+
+                let reply = proto::rpc::RpcSessionReply {
+                    session_result: Some(proto::rpc::rpc_session_reply::SessionResult::Rejected(true)),
+                };
+                framed.send(reply.to_encoded_bytes().into()).await?;
+                Err(RpcError::NegotiationClientNoSupportedVersion)
+            },
+            Ok(Some(Err(err))) => Err(err.into()),
+            Ok(None) => Err(RpcError::ClientClosed),
+            Err(_elapsed) => Err(RpcError::NegotiationTimedOut),
+        }
+    }
+}
+
+struct ActivePeerRpcService<TSvc, TSubstream> {
+    config: RpcServer,
+    node_id: NodeId,
+    service: TSvc,
+    framed: Option<CanonicalFraming<TSubstream>>,
+    shutdown_signal: OptionalShutdownSignal,
+}
+
+impl<TSvc, TSubstream> ActivePeerRpcService<TSvc, TSubstream>
```

Comment:
Nit: This guy needs a `fn new(...)` I think

Reply:
Mabe - it's private and basically a glorified internal function, so I'd say no.